docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+
+*   flexibility in scaling Trillian deployments,
+*   easier recovery from corrupt/failed database deployments since operators
+can simply delete it and allow it to rebuild itself from the commit log.
+
+Initially, this will be built using Apache Kafka for the commit log, with
+Apache HBase instances for the serving databases, since this is what Cloudflare
+has operational experience in running, but other distributed commit-log and
+database engines may be available.
+
+Having Trillian support a commit-log based storage system too will ensure
+Trillian doesn't inadvertently tie itself exclusively to strong globally
+consistent storage.
+
+## Background
+
+Trillian currently supports two storage technologies, mysql and Spanner, which
+provide strong global consistency.
+
+The design presented here requires:
+
+*   A durable, ordered, and immutable commit log.
+*   A "local" storage mechanism which can support the operations required by
+    the Trillian log_storage API.
+
+
+
+## Design Overview
+
+![overview diagram](commit_log_based_storage_design_overview.png "Overview")
+
+The `leaves` topic is the canonical source of truth for the ordering of leaves
+in a log.
+
+The `STHs` topic is a list of all STHs for a given log.
+
+Kafka topics are configured never to expire entries (this is a supported mode),
+and Kafka is known to scale to multiple terabytes within a single partition.
+
+HBase instances are one-per-cluster, built from the contents of the Kafka
+topics, and, consequently, are essentially disposable.
+
+Queued leaves are sent by the Trillian frontends to the Kafka `Leaves` topic.
+Since Kafka topics are append-only and immutable, this effectively sequences
+the entries in the queue.
+The current master sequencer consumes the `Leaves` topic, uses the Kafka sequence
+number of each dequeued leaf as its Trillian Log sequence number, and adds it to
+the local Merkle tree ensuring the sequence numbers match.
+
+When a new tree head is created, the sequencer writes this to the STHs topic
+before committing the tree update to the local database.
+
+Since the commit log forms the source of truth for the log entry ordering and
+committed STHs, everything else can be derived from that. This means that
+updates to the serving HBase DBs can be made to be idempotent, which means that
+the transactional requirements of Trillian's LogStorage APIs can be relaxed.
+
+The addition of this style of storage implementation requires that Trillian
+does not guarantee the perfect deduplication of entries, even though it may be
+possible to do so with some storage implementations.  i.e. personalities MUST
+present LeafIdentityHashes, and Trillian MAY deduplicate.
+
+## Detailed Design
+
+#### Enqueuing leaves
+
+RPC calls to frontend `QueueLeaves` results in the leaves being individually
+added to the Kafka topic `Leaves`.  They need to be added individually to allow
+the Kafka topic sequencing to be the definitive source of log sequence
+information.
```

Comment:
I'm not sure that's a good idea; the local DBs can soak up load, so operators are at liberty to scale up on demand (modulo replay lag), Kafka might be able to scale up for read requests too (I don't know enough about it to say), but e.g. other commit queue implementations might not.